### ISSUE TITLE
0x52 - Adversary can abuse hanging approvals left by PSwapLib.swap to bypass reward fees

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2023-07-blueberry-judging#104]
0x52

high

# Adversary can abuse hanging approvals left by PSwapLib.swap to bypass reward fees
## Summary

PSwapLib.swap can be manipulated to leave hanging allowances via the expectedRewards input. These can then be abused to swap reward tokens out of order allowing them to bypass fees.

## Vulnerability Detail

[AuraSpell.sol#L247-L257](https://github.com/sherlock-audit/2023-07-blueberry/blob/main/blueberry-core/contracts/spell/AuraSpell.sol#L247-L257)

            _doCutRewardsFee(sellToken);
            if (
                expectedRewards[i] != 0 &&
                !PSwapLib.swap(
                    augustusSwapper,
                    tokenTransferProxy,
                    sellToken,
                    expectedRewards[i],
                    swapDatas[i]
                )
            ) revert Errors.SWAP_FAILED(sellToken);

AuraSpell#closePositionFarm allows the user to specify any expectedRewards they wish. This allows the user to approve any amount, even if the amount is much larger than they would otherwise use. The can abuse these hanging approvals to swap tokens out of order and avoid paying reward fees.

Example:
Assume there are two rewards, token A and token B. Over time a user's position accumulates 100 rewards for each token. Normally the user would have to pay fees on those rewards. However they can bypass it by first creating hanging approvals. The user would start by redeeming a very small amount of LP and setting expectedRewards to uint256.max. They wouldn't sell the small amount leaving a very large approval left for both tokens. Now the user withdraws the rest of their position. This time they specify the swap data to swap token B first. The user still has to pay fees on token A but now they have traded token B before any fees can be taken on it. 

## Impact

User can bypass reward fees

## Code Snippet

[AuraSpell.sol#L184-L286](https://github.com/sherlock-audit/2023-07-blueberry/blob/main/blueberry-core/contracts/spell/AuraSpell.sol#L184-L286)

[ConvexSpell.sol#L179-L229](https://github.com/sherlock-audit/2023-07-blueberry/blob/main/blueberry-core/contracts/spell/ConvexSpell.sol#L179-L229)

## Tool used

Manual Review

## Recommendation

After the swap reset allowances to 0

**Comment by IAm0x52** [comment]
Escalate

This was wrongly excluded. The judging comments above are incorrect. The code allows the user to make any swap in whatever order they want. This is what allows the user to bypass the fees. The user is meant to swap in the order that the tokens are listed but by swapping in a different order they can bypass the cut. Assume we have two tokens to claim. Normally the flow would be:

RewardCut A > Swap A > RewardCut B > Swap B

Using the methodology I've shown above the user can instead:

RewardCut A > Swap B > RewardCutB > Swap A

Here we see that B can be swapped before the reward cut allowing the fee to be bypassed.

**Comment by sherlock-admin2** [comment]
> Escalate
> 
> This was wrongly excluded. The judging comments above are incorrect. The code allows the user to make any swap in whatever order they want. This is what allows the user to bypass the fees. The user is meant to swap in the order that the tokens are listed but by swapping in a different order they can bypass the cut. Assume we have two tokens to claim. Normally the flow would be:
> 
> RewardCut A > Swap A > RewardCut B > Swap B
> 
> Using the methodology I've shown above the user can instead:
> 
> RewardCut A > Swap B > RewardCutB > Swap A
> 
> Here we see that B can be swapped before the reward cut allowing the fee to be bypassed.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by Shogoki** [comment]
> Escalate
> 
> This was wrongly excluded. The judging comments above are incorrect. The code allows the user to make any swap in whatever order they want. This is what allows the user to bypass the fees. The user is meant to swap in the order that the tokens are listed but by swapping in a different order they can bypass the cut. Assume we have two tokens to claim. Normally the flow would be:
> 
> RewardCut A > Swap A > RewardCut B > Swap B
> 
> Using the methodology I've shown above the user can instead:
> 
> RewardCut A > Swap B > RewardCutB > Swap A
> 
> Here we see that B can be swapped before the reward cut allowing the fee to be bypassed.

Agree with escalation. As the user can specify any `expectedRewards` and any `callData` in any order, this is a valid attack vector.

**Comment by hrishibhat** [comment]
@Gornutz

**Comment by Gornutz** [comment]
The paraswap lib is already resetting the token approval to zero prior to allocating the allowance for each swap - https://github.com/sherlock-audit/2023-07-blueberry/blob/7c7e1c4a8f3012d1afd2e598b656010bb9127836/blueberry-core/contracts/libraries/Paraswap/PSwapLib.sol#L15

**Comment by Shogoki** [comment]
> The paraswap lib is already resetting the token approval to zero prior to allocating the allowance for each swap - https://github.com/sherlock-audit/2023-07-blueberry/blob/7c7e1c4a8f3012d1afd2e598b656010bb9127836/blueberry-core/contracts/libraries/Paraswap/PSwapLib.sol#L15

I do not think this helps preventing this issue. As the approval is resetted before the swap, there still can be a hanging approval.
Issue is that the approval is set/reset for the specified sellToken, but the user can swap any other token by specifying different calldatas. Therefore, i think the stated attack path is still possible

**Comment by Gornutz** [comment]
Gotcha, then yes seems valid

**Comment by hrishibhat** [comment]
Result:
Medium
Unique
This is a valid medium issue

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [IAm0x52](https://github.com/sherlock-audit/2023-07-blueberry-judging/issues/104/#issuecomment-1694746004): accepted
